### PR TITLE
refactor: add xmemcpyz() and use it in place of some xstrlcpy()

### PR DESCRIPTION
**Pull request by zeertzjq**
Problem:  Using xstrlcpy() when the exact length of the string to be
          copied is known is not ideal because it requires adding 1 to
          the length and an unnecessary strlen().
Solution: Add xmemcpyz() and use it in place of such xstrlcpy() calls.

Related #859

### CODE REVIEW

File: src/nvim/memory.c [hunk 1]
```diff
@@ -222,6 +222,21 @@ void *xmemdupz(const void *data, size_t len)
   return memcpy(xmallocz(len), data, len);
 }
 
+/// Duplicates `len` bytes of `src` to `dst` and zero terminates it.
+/// and returns a pointer to the allocated memory. If the allocation fails,
+/// the program dies.
```
Comment (justinmk) [inline src/nvim/memory.c:227]: Is this actually true? I see that this docstring was copied from `xmemdupz`, which calls `xmallocz` which will terminate the program. But this new function does not call anything that would kill if allocation fails.

Comment (zeertzjq) [reply src/nvim/memory.c:227]: Oops, forgot to delete these two lines.

**Comment by justinmk** [comment]
coverity warning:

```
*** CID 492655:  Memory - corruptions  (OVERRUN)
/src/nvim/cmdexpand.c: 3050 in expand_shellcmd()
3044         }
3045     
3046         size_t l = (size_t)(e - s);
3047         if (l > MAXPATHL - 5) {
3048           break;
3049         }
>>>     CID 492655:  Memory - corruptions  (OVERRUN)
>>>     Overrunning buffer pointed to by "s" of 1 bytes by passing it to a function which accesses it at byte offset 4090 using argument "l" (which evaluates to 4091).
3050         expand_shellcmd_onedir(buf, s, l, pat, matches, numMatches, flags, &found_ht, &ga);
3051         if (*e != NUL) {
3052           e++;
3053         }
3054       }
3055       *matches = ga.ga_data;
```

**Comment by zeertzjq** [comment]
It's already fixed.

**Pull request merged**